stream_consumer.go

```diff
@@ -69,12 +110,41 @@ type StreamEndpointConfigOpt func(config *StreamEndpointConfig)
 
 type EndpointType uint8
 
+// Add options for the stream endpoint creation, this can be used when stream endpoints are created under the hood by the methods below.
+func WithStreamEndpointOptions(opts ...StreamEndpointConfigOpt) Option {
+	return Option{Opt: func(gaz *Gaz) error {
+		gaz.streamEndpointOptions = opts
+		return nil
+	}}
+}
+
+// Call this method to create a stream consumer with the full stream name (pattern: "serviceName.streamName")
+// The service name is resolved via service discovery
+// Under the hood we make sure that only 1 subscription is done for a service, even if multiple streams are created on the same service
+func (g *Gaz) DiscoverAndConsumeStream(fullStreamName string, opts ...ConsumerConfigOpt) (StreamConsumer, error) {
+	srv, stream := ParseStreamName(fullStreamName)
+	return g.DiscoverAndConsumeServiceStream(srv, stream, opts...)
+}
+
+// Call this method to create a stream consumer
+// The service name is resolved via service discovery
+// Under the hood we make sure that only 1 subscription is done for a service, even if multiple streams are created on the same service
+func (g *Gaz) DiscoverAndConsumeServiceStream(service, stream string, opts ...ConsumerConfigOpt) (StreamConsumer, error) {
+	return g.createConsumer([]string{SdPrefix + service}, stream, opts...)
+}
+
+// Call this method to create a stream consumer with the service endpoints and the stream name
+// Under the hood we make sure that only 1 subscription is done for a service, even if multiple streams are created on the same service
+func (g *Gaz) ConsumeStream(endpoints []string, stream string, opts ...ConsumerConfigOpt) (StreamConsumer, error) {
```

Comment:
So, this one does not consume from a service, it can't be used with Service Discovery?
I would like to be sure we won't have to change too much the code and use different concepts if we use SD or not.
In both cases, we have services. I would really love to only see some changes with the gRPC balancer and have the rest of the code the same

Reply:
Yes this one is not using service discovery.
The problem we have is that with service discovery, 1 service name yields n endpoints that we can use with the load balancer; but if we want to do load balancing without service discovery, we need to pass multiple endpoints...

I don't think it would be so nice to have an API for service discovery where we can pass multiple service names as arguments, that's why I created this additional method.